js-exercises/animal-sort/animalSort.test.js

```diff
@@ -0,0 +1,47 @@
+import { animalSort } from './animalSort';
+
+describe('animalSort', () => {
+  it('should return empty array if empty array is passed in', () => {
+    expect(() => animalSort([{}])).toThrowError(/[A-Za-z0-9]+/g);
+    expect(() => animalSort([{ name: '', legs: '' }])).toThrowError(/[A-Za-z0-9]+/g);
+    expect(() => animalSort({})).toThrowError(/[A-Za-z0-9]+/g);
+  });
+
+  it('should return empty array if empty array is passed in', () => {
+    expect(animalSort([])).toEqual([]);
+  });
+
+  it('should return a sorted array of animal objects by their number of legs', () => {
+    const arr = [
+      { name: 'Dog', numberOfLegs: 4 },
+      { name: 'Bird', numberOfLegs: 2 },
+      { name: 'Snake', numberOfLegs: 0 },
+    ];
+    const result = [
+      { name: 'Snake', numberOfLegs: 0 },
+      { name: 'Bird', numberOfLegs: 2 },
+      { name: 'Dog', numberOfLegs: 4 },
+    ];
+    expect(animalSort(arr)).toEqual(result);
+  });
+
+  it('should also return sorted names of animals with same number of legs', () => {
+    const arr = [
+      { name: 'Cat', numberOfLegs: 4 },
+      { name: 'Snake', numberOfLegs: 0 },
+      { name: 'Dog', numberOfLegs: 4 },
+      { name: 'Pig', numberOfLegs: 4 },
+      { name: 'Human', numberOfLegs: 2 },
+      { name: 'Bird', numberOfLegs: 2 },
+    ];
+    const result = [
+      { name: 'Snake', numberOfLegs: 0 },
+      { name: 'Bird', numberOfLegs: 2 },
+      { name: 'Human', numberOfLegs: 2 },
+      { name: 'Cat', numberOfLegs: 4 },
+      { name: 'Dog', numberOfLegs: 4 },
+      { name: 'Pig', numberOfLegs: 4 },
+    ];
+    expect(animalSort(arr)).toEqual(result);
```

Comment:
Here's 1 more test for you.
Add it, and see if your current program can do that?

```suggestion
    expect(animalSort(arr)).toEqual(result);
    
    const arr2 = arr.map(object => object.numberOfLegs = null);
    const result2 = [
    { name: 'Bird', numberOfLegs: null},
    { name: 'Cat', numberOfLegs: null},
    { name: 'Dog', numberOfLegs: null},
    { name: 'Human', numberOfLegs: null},
    { name: 'Pig', numberOfLegs: null},
    { name: 'Snake', numberOfLegs: null},
    ];
    
    expect(animalSort(arr2)).toEqual(result2);
```

Reply:
yes, it worked. Because, in the above test case number of legs is null. So it will check for the condition
```
if (animalA.numberOfLegs === animalB.numberOfLegs) {
  return animalA.name.localeCompare(animalB.name);
}
```
`null === null `gives true. So will do the localeCompare and return the same.

Reply:
Ahh right, and what if I keep some of them null and some of them number?
Will the result predictable in that case? And how will program tackle that situation?